jwql/instrument_monitors/pipeline_tools.py

```diff
@@ -0,0 +1,249 @@
+        # MIRI is limited to one frame per group
+        steps.remove('group_scale')
+    else:
+        steps = ['group_scale', 'dq_init', 'saturation', 'ipc', 'superbias', 'refpix', 'linearity',
+                 'persistence', 'dark_current', 'jump', 'rate']
+
+        # No persistence correction for NIRSpec
+        if instrument == 'NIRSPEC':
+            steps.remove('persistence')
+        else:
+            # NIRCam, NISISS, FGS all do not need group scale as nframes is
+            # always a multiple of 2
+            steps.remove('group_scale')
+
+    # IPC correction currently not done for any instrument
+    steps.remove('ipc')
+
+    # Initialize using PIPE_KEYWORDS so the steps will be in the right
+    # order
+    req = OrderedDict({})
+    for key in steps:
+        req[key] = True
+    for key in PIPE_KEYWORDS.values():
+        if key not in req.keys():
+            req[key] = False
+
+    return req
+
+
+def image_stack(file_list):
+    """Given a list of fits files containing 2D images, read in all data
+    and place into a 3D stack
+
+    Parameters
+    ----------
+    file_list : list
+        List of fits file names
+
+    Returns
+    -------
+    cube : numpy.ndarray
+        3D stack of the 2D images
+    """
+    exptimes = []
+    for i, input_file in enumerate(file_list):
+        with fits.open(input_file) as hdu:
+            image = hdu[1].data
+            exptime = hdu[0].header['EFFINTTM']
+            num_ints = hdu[0].header['NINTS']
+
+        # Stack all inputs together into a single 3D image cube
+        if i == 0:
+            ndim_base = image.shape
+            if len(ndim_base) == 3:
+                cube = copy.deepcopy(image)
+            elif len(ndim_base) == 2:
+                cube = np.expand_dims(image, 0)
+        else:
+            ndim = image.shape
+            if ndim_base[-2:] == ndim[-2:]:
+                if len(ndim) == 2:
+                    image = np.expand_dims(image, 0)
+                elif len(ndim) > 3:
+                    raise ValueError("4-dimensional input slope images not supported.")
+                cube = np.vstack((cube, image))
+            else:
+                raise ValueError("Input images are of inconsistent size in x/y dimension.")
+        exptimes.append([exptime] * num_ints)
+    return cube, exptimes
+
+
+def run_calwebb_detector1_steps(input_file, steps):
+    """Run the steps of calwebb_detector1 specified in the steps
+    dictionary on the input file
+
+    Parameters
+    ----------
+    input_file : str
+        File on which to run the pipeline steps
+
+    steps : collections.OrderedDict
+        Keys are the individual pipeline steps (as seen in the
+        PIPE_KEYWORDS values above). Boolean values indicate
+        whether a step should be run or not. Steps are run in the
+        official calwebb_detector1 order.
+    """
+    first_step_to_be_run = True
+    for step_name in steps.keys():
+        if steps[step_name]:
```

Comment:
@bhilbert4 Not a big deal here, but this could also be written as:

```python
for step_name in steps:
    if steps[step_name]:
```

It is slightly cleaner and doesn't require the `.keys()` method.

Reply:
Stripped off `keys()`.